Change `find_accessor` to only emit identifiers

andreaphylum opened the pull request:
This PR changes the `AccessGraph::find_accessor` method to only emit `identifier` nodes as accessors, continuing the search upwards into the tree if that is not possible. 

Tests are added to validate that the destructuring assignments work as intended. A failing, ignored test is also added to document that we currently don't handle shorthand identifiers correctly.

Closes #48.

andreaphylum commented on vuln-reach/src/javascript/lang/accesses.rs:
```diff
@@ -165,17 +165,23 @@ impl<'a> AccessGraph<'a> {
+                // Find the next parent identifier if node is an anonymous declaration.
                 "class" | "function" | "arrow_function" => {
-                    return Self::find_accessor(cursor_cache, parent);
+                    cur_depth = depth;
+                    node = parent;
```
I don't like this part in this form because its effects ("restarting" the computation) are not super evident compared to recursing. Happy to accept suggestions here.

cd-work replied:
I'm fine with it tbh. Though I didn't really mind the recursion either (and it's probably *simpler* imo).

andreaphylum replied:
The recursion is definitely easier to reason about, at least. My concern is more with the fact that this is on a hot path and Rust has no tail recursion and I'm not sure that can be inlined away. Then again I ran the benchmarks and there is so much variance as to make any assessment inconsistent, so it would probably be fine to stick with recursion for clarity's sake.

cd-work commented on vuln-reach/src/javascript/lang/accesses.rs:
```diff
@@ -165,17 +165,23 @@ impl<'a> AccessGraph<'a> {
+        mut node: Node<'a>,
+    ) -> Option<Node<'a>> {
         let cursor = cursor_cache.cursor(node).unwrap();
+        let mut cur_depth = 0;
```
I think this is a bad name. Current depth is `depth` itself. This is more akin to the origin/target of the function.

cd-work commented on vuln-reach/src/javascript/lang/accesses.rs:
```diff
@@ -188,17 +194,28 @@ impl<'a> AccessGraph<'a> {
                         // Find next parent identifier if node is in LHS of the assignment.
-                        return Self::find_accessor(cursor_cache, parent);
+                        cur_depth = depth;
+                        node = parent;
```
Node is never actually used for anything other than constructing the cursor I think? So it should be fine removing this statement too?

It kinda makes sense since this loop inherently moves through the parents so a no-op is basically the same as recursion (except for `class_declaration` with `depth == 0`).

andreaphylum replied:
I did only reassign depth at first, but then I noticed that `&node.start_byte()` is actually used, right on top of this snippet. So a proper recursion would entail reassigning the node as well. 

I'm not super clear on where of those branches we want to have a proper recursion/"reinitialization"; one could argue that just keeping going with the loop all the way through could be ok too. Could we maybe make a special case of anonymous declarations without using `depth`?

andreaphylum commented:
I changed `find_accessor` so that it now compares the node itself rather than the depth.

I also added a couple more granular unit tests for accessors. From those, it doesn't look like resetting the state would be helpful. 

In the current state, the recursion happens in these two cases only:
- From anonymous `class`, `function` or `arrow_function` nodes:
  ```js
  function foo() {
    () => { bar }
    class Bar { constructor() { bar } }
    function baz() { bar }
  }
  ```
  ```
  function_declaration [0, 0] - [4, 1]
    name: identifier [0, 9] - [0, 12]
    parameters: formal_parameters [0, 12] - [0, 14]
    body: statement_block [0, 15] - [4, 1]
      expression_statement [1, 2] - [1, 15]
        arrow_function [1, 2] - [1, 15]
          parameters: formal_parameters [1, 2] - [1, 4]
          body: statement_block [1, 8] - [1, 15]
            expression_statement [1, 10] - [1, 13]
              identifier [1, 10] - [1, 13]
      class_declaration [2, 2] - [2, 37]
        name: identifier [2, 8] - [2, 11]
        body: class_body [2, 12] - [2, 37]
          member: method_definition [2, 14] - [2, 35]
            name: property_identifier [2, 14] - [2, 25]
            parameters: formal_parameters [2, 25] - [2, 27]
            body: statement_block [2, 28] - [2, 35]
              expression_statement [2, 30] - [2, 33]
                identifier [2, 30] - [2, 33]
      function_declaration [3, 2] - [3, 24]
        name: identifier [3, 11] - [3, 14]
        parameters: formal_parameters [3, 14] - [3, 16]
        body: statement_block [3, 17] - [3, 24]
          expression_statement [3, 19] - [3, 22]
            identifier [3, 19] - [3, 22]
  ```
- When the identifier is in the LHS of an assignment or variable/lexical declaration.
  ```js
  function foo() {
    let bar = baz
    bar = baz
    bar += baz
  }
  ```
  ```
  function_declaration [0, 0] - [4, 1]
    name: identifier [0, 9] - [0, 12]
    parameters: formal_parameters [0, 12] - [0, 14]
    body: statement_block [0, 15] - [4, 1]
      lexical_declaration [1, 2] - [1, 15]
        variable_declarator [1, 6] - [1, 15]
          name: identifier [1, 6] - [1, 9]
          value: identifier [1, 12] - [1, 15]
      expression_statement [2, 2] - [2, 11]
        assignment_expression [2, 2] - [2, 11]
          left: identifier [2, 2] - [2, 5]
          right: identifier [2, 8] - [2, 11]
      expression_statement [3, 2] - [3, 12]
        augmented_assignment_expression [3, 2] - [3, 12]
          left: identifier [3, 2] - [3, 5]
          right: identifier [3, 9] - [3, 12]
  ```

If we were to restart the computation from the parent node of these identifiers, in none of those cases we would be starting from the `name` field of a named declaration, and hence we would never incur in a case where `depth == 0`  on a named declaration. This means that recursing and continuing the iteration has no practical difference.

Another simpler way of seeing this: `identifier` nodes are _always_ leaves and so, since the recursion moves through parents, they could only ever appear as `node` argument in the outermost call.

@cd-work, do you know if this is incorrect or I'm missing something here?

cd-work commented on vuln-reach/src/javascript/lang/accesses.rs:
```diff
@@ -168,15 +168,14 @@ impl<'a> AccessGraph<'a> {
+                "class_declaration" | "function_declaration" => {
+                    return parent.child_by_field_name("name").filter(|&name| name != node);
                 },
+                // Find the next parent identifier if node is an anonymous declaration.
+                "class" | "function" | "arrow_function" => (),
```
No need to special case this from `_`.

cd-work commented on vuln-reach/src/javascript/lang/accesses.rs:
```diff
@@ -186,19 +185,26 @@ impl<'a> AccessGraph<'a> {
-                        // Find next parent identifier if node is in LHS of the assignment.
-                        return Self::find_accessor(cursor_cache, parent);
-                    } else {
+                    if !lhs.byte_range().contains(&node.start_byte()) {
```
This is fine since any previously visited child node works here. We just need to know if we came from lhs or rhs.

cd-work commented on vuln-reach/src/javascript/lang/accesses.rs:
```diff
@@ -168,15 +168,14 @@ impl<'a> AccessGraph<'a> {
-                    return Self::find_accessor(cursor_cache, parent);
+                // Check node to avoid declarations using themselves as accessor.
+                "class_declaration" | "function_declaration" => {
+                    return parent.child_by_field_name("name").filter(|&name| name != node);
```
This **will** return `None`, which is incorrect.
